File: main/hw1-insertion-sort/hw1-insertion-sort.java
import java.util.Scanner;

final class Solution {
    public static void main(final String[] args) {
        final int[] a = readRecord();
        Algo.insertionSort(a);
        print(a);
    }

    private static int[] readRecord() {
        try (final Scanner sc = new Scanner(System.in)) {
            final int[] a = new int[sc.nextInt()];
            for (int i = 0; i != a.length; ++i) a[i] = sc.nextInt();
            return a;
        }
    }

    private static void print(final int[] a) {
        System.out.print("[");

        String sep = "";
        for (final int x : a) {
            System.out.format("%s%d", sep, x);
            sep = ",";
        }

        System.out.println("]");
    }

    private Solution() { throw new AssertionError(); }
}

final class Algo {
    static void insertionSort(final int[] a) {
        if (a.length < 2) return;

        for (int right = 1; right != a.length; ++right) {
            final int elem = a[right];

            int left = right;
            for (; left != 0 && elem < a[left - 1]; --left)
                a[left] = a[left - 1];

            a[left] = elem;
        }
    }

    private Algo() { throw new AssertionError(); }
}

File: main/hw1-insertion-sort/hw1-insertion-sort.cpp
#include <iostream>
#include <iterator>
#include <utility>
#include <vector>

namespace {
    template<typename It>
    constexpr void insertion_sort(const It first, const It last) noexcept
    {
        if (first == last) return;

        for (auto right = first + 1; right != last; ++right) {
            auto elem = std::move(*right);

            auto left = right;
            for (; left != first && elem < left[-1]; --left)
                *left = std::move(left[-1]);

            *left = std::move(elem);
        }
    }

    std::vector<int> read_record()
    {
        std::vector<int>::size_type n {};
        std::cin >> n;

        std::vector<int> a (n);
        for (auto& x : a) std::cin >> x;
        return a;
    }

    void print(const std::vector<int>& a)
    {
        std::cout << '[';

        auto sep = "";
        for (const auto x : a) {
            std::cout << sep << x;
            sep = ",";
        }

        std::cout << "]\n";
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    auto a = read_record();
    insertion_sort(begin(a), end(a));
    print(a);
}
